### PR TITLE
linux_rpi5: init at 6.1.63-stable_20231123

### PR DESCRIPTION
**Pull request by leo60228**
## Description of changes

Builds the BCM2712/Raspberry Pi 5 defconfig for linux-rpi. linux_rpi4 works, but uses 4K pages instead of 16K, and is thus not recommended by Raspberry Pi.

Relates to #260754.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by IogaMaster — APPROVED** [review]

**Comment by alyssais** [comment]
There's this note in linux-kernels.nix about not adding new vendor kernels:

>     # NOTE: PLEASE DO NOT ADD NEW VENDOR KERNELS TO NIXPKGS.
>     # New vendor kernels should go to nixos-hardware instead.
>     # e.g. https://github.com/NixOS/nixos-hardware/tree/master/microsoft/surface/kernel

**Comment by leo60228** [comment]
Technically, this isn't a *new* kernel, just building an existing one with
a different config....

Adding it to nixos-hardware isn't an amazing option, since compiling the
kernel on-device takes several hours.

**Comment by alyssais** [comment]
It would be good indeed if nixos-hardware could be built by Hydra.

**Comment by malteneuss** [comment]
Is there anything missing for this PR to be merged?

@leo60228 What is the relation to https://gitlab.com/vriska/nix-rpi5/-/blob/main/linux-rpi.nix?ref_type=heads from your wiki page https://wiki.nixos.org/wiki/NixOS_on_ARM/Raspberry_Pi_5 ? It looks equivalent, so could we remove that external dependency?

**Comment by leo60228** [comment]
> Is there anything missing for this PR to be merged?

Policy clarification: this probably *should* go in nixos-hardware, but that's not built by Hydra.

> What is the relation to https://gitlab.com/vriska/nix-rpi5/-/blob/main/linux-rpi.nix?ref_type=heads from your wiki page https://wiki.nixos.org/wiki/NixOS_on_ARM/Raspberry_Pi_5 ? It looks equivalent, so could we remove that external dependency?

This PR is an attempt to upstream that, yes.

**Comment by malteneuss** [comment]
Ah nice. Who could we ping for clarification? You're argument of not actually adding a new config seems reasonable.

**Comment by malteneuss** [comment]
I just tried compiling the rpi5 kernel natively via bin-fmt/qemu and gave up after 4h on an average laptop. So if there's no cached binary, the user experience indeed goes towards zero :(
Is this another argument to merge it here and not nixos-hardware?

**Comment by alyssais** [comment]
Well, I wrote that note, so I know what I meant by it.  Other people generally involved in this sort of conversation include @samueldr, @k900, @RaitoBezarius.  (I think we possibly even discussed this PR on Matrix at the time.)

**Comment by alyssais** [comment]
> Is this another argument to merge it here and not nixos-hardware?

It's an argument for having Hydra on nixos-hardware.

**Review by samueldr — CHANGES_REQUESTED** [review]
We're not making the mistake we kept making in the past with this new hostile vendor platform.

If anything has to be done, a PR removing `linuxPackages_rpi*` would be the more proper action to take.

**Comment by malteneuss** [comment]
@leo60228 Could you create a PR against Nixos-hardware then? Letting it be build by Hydra could be another topic/PR.

@samueldr What are your thoughts on Hydra for Nixos-hardware? And is there any deprecaction workflow we could start here?

**Comment by samueldr** [comment]
> @samueldr What are your thoughts on Hydra for Nixos-hardware? And is there any deprecaction workflow we could start here?

Due to how infrastructure stuff is going at the moment, it's not the time to plan this. So for now the status quo of not changing this is the better option.

Though, we might want to already start planning to move rpi1/rpi2 to nixos-hardware and deprecate it "outright" considering anyway those are not being built in any form by our infra.

It would leave the 64-bit pair as a whole, and those one we would need to figure out the plan going forward.

TLDR: there is no pressure to remove anything at the moment. I mainly wanted to confirm @alyssais's understanding, and signal the intention clearly.

**Comment by leo60228** [comment]
Superseded by https://github.com/NixOS/nixos-hardware/pull/927.
